### DISCUSSION TITLE
Skipping vite build. Fails on Rails 7 with Capistrano 3

### DISCUSSION THREAD

**Discussion by octave** [ElMassimo/vite_ruby#186] [Q&A] [answered]
Using:

- `vite_rails: 3.0.5`
- `rails: 7.0.1`
- `capistrano: 3.19.0`

After deploying when making changes to assets, it works fine. But when deploying without making any changes to the assets then during `rake assets:precompile`, what happens is `Skipping vite build. Watched files have not changed since the last build`. But after that the assets are not linked properly and the rails app crash: `ActionView::Template::Error (Vite Ruby can't find entrypoints/application.js in public/vite/manifest.json or public/vite/manifest-assets.json.`

**Comment by ElMassimo** [comment, marked as answer]
__UPDATE:__ Since [Vite Ruby 3.1](https://github.com/ElMassimo/vite_ruby/pull/200), it's now easier to configure `capistrano-rails`.

[See the documentation](https://vite-ruby.netlify.app/guide/deployment.html#using-capistrano).

---

Seems like a capistrano configuration problem.

The build will only be skipped if a [temporary file](https://github.com/ElMassimo/vite_ruby/blob/main/vite_ruby/lib/vite_ruby/builder.rb#L40) from a previous build is found, and no files have changed since the last successful build.

Normally, if that file persists, then the public dir along with the Vite output dir would also persist.

Make sure that the public dir is not being cleared upon deployment to avoid rebuilding unnecessarily, or force a build on every deployment (less efficient).

**Reply by antoinematyja** [reply]
I had the same issue and can confirm this.

In my case, the vite build task was checking in `tmp/cache/vite/last-build-<env>.json`. Since `tmp/cache` was symlinked by capistrano, the file from previous build was found and vite build was skipped.

But the build files were located in `public/vite`, which was not symlinked so could not be found in the last release. The solution is to add `public/vite` to capistrano `linked_dirs` (if you were using webpacker before, this replaces `public/packs`).

Note that if you already deployed without symlinking `public/vite`, you will have to manually move build files to the shared directory, or delete the `last-build-<env>.json` file and re-deploy to get build files in the shared directory.

**Reply by ElMassimo** [reply]
Hi Antoine, thanks for providing details about why this problem may occur.

I think it would be useful to have a _Capistrano_ sub-section in the [_Deployment_](https://vite-ruby.netlify.app/guide/deployment.html) section, documenting which files should be linked (such as [this one for webpacker](https://makandracards.com/makandra/491354-configuring-webpacker-deployments-with-capistrano)).

I'm not using Capistrano for deployments, so I'd appreciate contributions to make sure the instructions stay relevant and up-to-date.

**Reply by antoinematyja** [reply]
I had the same thought, if I have some spare time I will look into this 🙂

I found 2 other caveats trying to update my rails/vue app from webpacker to vite_rails:

- When the `vite build` task is run, the `public/vite` directory is emptied, then the assets subdirectory is built and the manifest.json is created. This is different from webpacker which keeps built files from previous deployments.
If the `public/vite` directory is symlinked, the previous assets are deleted in the middle of the deployment. But these files are still required for the previous release deployed (which is still served until the new deployment is finished). Is there a way to have previous `public/vite` content coexist with the one from a new release? Seems complicated since we should merge json manifests...
If it's not possible, I think we must force a new vite build on each deployment.

- The 2nd caveat is not related directly to this package but only showed up when using it.
When I created my app with webpacker + Vue, this sample code was added:
https://github.com/rails/webpacker/blob/5-x-stable/lib/install/examples/vue/hello_vue.js#L11
Vue is initialized after the `DOMContentLoaded` event, but with [vite legacy plugin](https://github.com/vitejs/vite/tree/main/packages/plugin-legacy), the js code is loaded asynchronously after this event, so Vue is never initialized.
I had to add a `<div id="app"></div>` to my index.html file and mount Vue on it to avoid awaiting for the event.
Again this caveat is not directly related to this project but I put it here in case it helps.

**Reply by ElMassimo** [reply]
Excellent, thank you!

Regarding the first point, [`emptyOutDir` is now `false` in production](https://github.com/ElMassimo/vite_ruby/issues/177), so the only remaining problem would be the replacement of the manifest files, which might be ok since they are cached in memory once they are first read (also, there's no way to safely merge them since the keys do not change between deployments but the compiled files would).

Linking the assets dir might not be desirable. Ideally, the public dir is treated as part of the app code and separate releases keep a separate copy of it (otherwise rollbacks are also not reliable).

**Reply by antoinematyja** [reply]
Oh ok, I see that someone had the same issue as me 😅 But the fix didn't apply to me because I'm currently setting things up on my staging environment. Maybe this should apply to non-dev environments instead of production-only? Otherwise I think you're right, the manifest should be read & cached on server start/restart so no issue here.

Linking the assets directory is appreciable to avoid building the front app again when deploying backend-only changes.

**Reply by ElMassimo** [reply]
> Maybe this should apply to non-dev environments instead of production-only?

Building assets in staging will still set the Vite build mode to [`production`](https://github.com/ElMassimo/vite_ruby/pull/180/files#diff-1ddc243ec52640a459e0fb6f93ec25ed160d1406139d8fc22b3da3522f48aa3dR44).

> Linking the assets directory is appreciable to avoid building the front app again when deploying backend-only changes.

An alternative would be to copy the public dir from the previous release when creating a new one, before the `assets:precompile` task is run by Capistrano. That would avoid affecting assets from previous releases (safer rollbacks), and doesn't require full recompilation if frontend assets did not change.

**Reply by antoinematyja** [reply]
Mmh not sure about this.

By default, when I run `RAILS_ENV=production rails vite:build`, the directory is not emptied, but with `RAILS_ENV=staging rails vite:build` it is. I added `build: { emptyOutDir: false }` to my `vite.config.js` and it did changed the behavior with `RAILS_ENV=staging`.

When I run the `vite:build` task in staging, vite is `building for staging`:

```
% RAILS_ENV=staging rails vite:build
Building with Vite ⚡️
vite v2.8.6 building for staging...
```

I guess your `config.mode` variable gets the `RAILS_ENV` value too.

**Reply by ElMassimo** [reply]
Thanks for testing it, happy to take a PR to change that to `emptyOutDir: isDevelopment`. I'll make sure to release a fix for that next week.

**Reply by ElMassimo** [reply]
Released the fix by @antoinematyja in [`vite-plugin-ruby@3.0.9`](https://github.com/ElMassimo/vite_ruby/blob/main/vite-plugin-ruby/CHANGELOG.md?rgh-link-date=2022-04-01T17%3A48%3A22Z#309-2022-04-01) 😃

**Reply by ElMassimo** [reply]
@antoinematyja Thanks for your feedback, released `vite_ruby@3.1.0` which should make it easier to configure Capistrano to work well with Vite Ruby. See:

- https://github.com/ElMassimo/vite_ruby/pull/200
